src/infra/core/src/repos/dataset_repository_s3.rs

```diff
@@ -26,76 +25,105 @@ use crate::utils::s3_context::S3Context;
         self.s3_context.recursive_delete(dataset_key_prefix).await
     }
 
-    async fn move_bucket_items_on_dataset_rename(
+    async fn resolve_dataset_alias(
         &self,
-        old_alias: &DatasetAlias,
-        new_alias: &DatasetAlias,
-    ) -> Result<(), MoveBucketItemsOnRenameError> {
-        assert!(
-            !new_alias.is_multi_tenant(),
-            "Multi-tenancy is not yet supported by S3 repo"
-        );
-
-        let new_key_prefix = self.s3_context.get_key(&new_alias.dataset_name);
-        if self
-            .s3_context
-            .bucket_path_exists(new_key_prefix.as_str())
-            .await?
-        {
-            return Err(MoveBucketItemsOnRenameError::NameCollision(
-                NameCollisionError {
-                    alias: new_alias.clone(),
-                },
-            ));
+        dataset: &dyn Dataset,
+    ) -> Result<DatasetAlias, InternalError> {
+        match dataset.as_info_repo().get("alias").await {
+            Ok(bytes) => {
+                let dataset_alias_str = std::str::from_utf8(&bytes[..]).int_err()?.trim();
+                let dataset_alias = DatasetAlias::try_from(dataset_alias_str).int_err()?;
+                Ok(dataset_alias)
+            }
+            Err(e) => Err(e.int_err()),
         }
+    }
 
-        let old_key_prefix = self.s3_context.get_key(&old_alias.dataset_name);
-        self.s3_context
-            .recursive_move(old_key_prefix, new_key_prefix)
-            .await?;
+    async fn save_dataset_alias(
+        &self,
+        dataset: &dyn Dataset,
+        dataset_alias: DatasetAlias,
+    ) -> Result<(), InternalError> {
+        dataset
+            .as_info_repo()
+            .set("alias", dataset_alias.to_string().as_bytes())
+            .await
+            .int_err()?;
 
         Ok(())
     }
+
+    fn stream_datasets_if<'s>(
+        &'s self,
+        alias_filter: impl Fn(&DatasetAlias) -> bool + Send + 's,
+    ) -> DatasetHandleStream<'s> {
+        Box::pin(async_stream::try_stream! {
+            let folders_common_prefixes = self.s3_context.bucket_list_folders().await?;
+            for prefix in folders_common_prefixes {
+                let mut prefix = prefix.prefix.unwrap();
+                while prefix.ends_with('/') {
+                    prefix.pop();
+                }
+
+                if let Ok(id) = DatasetID::try_from(format!("did:odf:{}", prefix)) {
+                    let dataset = self.get_dataset_impl(&id).await?;
+                    let dataset_alias = self.resolve_dataset_alias(&dataset).await?;
+                    if alias_filter(&dataset_alias) {
+                        let hdl = DatasetHandle::new(id, dataset_alias);
+                        yield hdl;
+                    }
+
+                }
+            }
+        })
+    }
+
+    fn normalize_alias(&self, alias: &DatasetAlias) -> DatasetAlias {
+        if alias.is_multi_tenant() {
+            alias.clone()
```

Comment:
Nitpick: Would it be better to accept `alias` param as owned to avoid excessive cloning?

Reply:
No, that's not easy. I'm getting alias references from higher level APIs, like `create_dataset` and `resolve_dataset_ref`